File: 달리기 경주.java
import java.util.Map;
import java.util.HashMap;

class Solution {

    public String[] solution(String[] players, String[] callings) {

        final Map<String, Integer> playerRanks = new HashMap<>();
        final Map<Integer, String> ranks = new HashMap<>();

        for (int i=0; i<players.length; i++) {

            playerRanks.put(players[i], i);
            ranks.put(i, players[i]);
        }

        for (final String calling: callings) {

            final int playerRank = playerRanks.get(calling);
            final String anotherPlayer = ranks.get(playerRank-1);

            playerRanks.put(calling, playerRank-1);
            playerRanks.put(anotherPlayer, playerRank);

            ranks.put(playerRank-1, calling);
            ranks.put(playerRank, anotherPlayer);

            players[playerRank-1] = calling;
            players[playerRank] = anotherPlayer;
        }

        return players;
    }
}

File: 달리기 경주.kt
class Solution {

    fun solution(players: Array<String>, callings: Array<String>): Array<String> {

        val playerRank = mutableMapOf<String, Int>()
        val rankPlayer = mutableMapOf<Int, String>()

        players
        .forEachIndexed{
            i, p ->
            playerRank[p] = i
            rankPlayer[i] = p
        }

        callings
        .forEach{
            
            val rank = playerRank[it]!!
            val other = rankPlayer[rank-1]!!

            playerRank[it] = rank-1
            playerRank[other] = rank
            rankPlayer[rank] = other
            rankPlayer[rank-1] = it
        }

        return playerRank
        .toList()
        .sortedBy{ (k,v) -> v }
        .map{ (k,v) -> k }
        .toTypedArray()
    }
}
